File: 01 - Disciplinas/Estrutura de Dados/Codigo/filaEst.cpp
/********************************************
* Exemplo de Fila com estrutura estatica    *
* filaEst.c                                 *
*********************************************/
#include <iostream>
using namespace std;
#define MAX 50

typedef int TIPOCHAVE;

struct REGISTRO{
    TIPOCHAVE chave;
};

struct FILA{
    REGISTRO A[MAX];
    int inicio;
    int nroElem;
};

void inicializarFila(FILA *f){
    f->inicio = 0;
    f->nroElem = 0;
}

int tamanhoFila(FILA *f) {
    return f->nroElem;
}

void listarFila (FILA *f){
    cout << "Fila: [ ";
    int i = f->inicio;
    int temp;
    for(temp = 0; temp < f->nroElem; temp++){
        cout << f->A[i].chave << " ";
        i = (i + 1) % MAX; // se o i passar do máximo, i volta a zero
    }
    cout << "]" << endl;
}

bool inserirElemFila(FILA *f, REGISTRO reg) {
   if (f->nroElem >= MAX) return false;
   int posicao = (f->inicio + f->nroElem) % MAX;
   f->A[posicao] = reg;
   f->nroElem++;
   return true; 
}

bool excluirElementoFila(FILA *f, REGISTRO *reg){
    if (f->nroElem == 0) return false;
    *reg = f->A[f->inicio];
    f->inicio = (f->inicio+1) % MAX;
    f->nroElem--;
    return true;
}

void reinicializaFila(FILA *f){
    inicializarFila(f);
}

int main() {
    return 0;
}

File: 01 - Disciplinas/Estrutura de Dados/Codigo/FilaEst.java
import javax.swing.JTextArea;

class Elemento {
    int chave;
    String nome;

    Elemento(int chave, String nome) {
        this.chave = chave;
        this.nome = nome;
    }
}

class FilaEstatica {
    int inicio;
    int numElem;
    int max;
    Elemento[] fila;

    FilaEstatica(int max) {
        fila = new Elemento[max];
        inicio = 0;
        numElem = 0;
        this.max = max;
    }

    int tamanho() {
        return numElem;
    }

    boolean inserirElemento(Elemento l) {
        if (numElem >= max)
            return false;
        int posicao = (inicio + numElem) % max;
        fila[posicao] = l;
        numElem++;
        return true;
    }

    Elemento excluirElemento() {
        if (numElem == 0)
            return null;
        Elemento l = fila[inicio];
        inicio = (inicio + 1) % max;
        numElem--;
        return l;

    }

    void imprime(JTextArea txtF) {
        txtF.append("\nListagem da Fila\n");
        int pos = inicio;
        for (int i = 0; i < numElem; i++) {
            // System.out.println("chave " + fila[i].chave + " nome " + fila[i].nome);
            txtF.append("chave " + fila[pos].chave + " nome " + fila[pos].nome + "\n");
            pos = (pos + 1) % max;
        }
        txtF.append("------------------------------------\n");
    }

    void reinicializaFila() {
        inicio = 0;
        numElem = 0;
    }
}
